server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -0,0 +1,313 @@
+                    }
+                } catch (final Exception e) {
+                    tragedy.setTragicException(e);
+                    closeOnTragicEvent(e);
+                    throw e;
+                }
+            }
+            // Do we need remote writes in sync fashion ?
+            // If we don't , we should swallow FileAlreadyExistsException while writing to remote store
+            // and also verify for same during primary-primary relocation
+            // Writing remote in sync fashion doesn't hurt as global ckp update
+            // is not updated in remote translog except in primary to primary recovery.
+            if (generation == null) {
+                if (closed.get() == false) {
+                    return upload(primaryTerm, current.getGeneration() - 1);
+                } else {
+                    return upload(primaryTerm, current.getGeneration());
+                }
+            } else {
+                return upload(primaryTerm, generation);
+            }
+        }
+    }
+
+    private boolean upload(Long primaryTerm, Long generation) throws IOException {
+        logger.trace("uploading translog for {} {} ", primaryTerm, generation);
+        try (
+            TranslogCheckpointTransferSnapshot transferSnapshotProvider = new TranslogCheckpointTransferSnapshot.Builder(
+                primaryTerm,
+                generation,
+                location,
+                readers,
+                Translog::getCommitCheckpointFileName
+            ).build()
+        ) {
+            Releasable transferReleasable = Releasables.wrap(deletionPolicy.acquireTranslogGen(getMinFileGeneration()));
+            return translogTransferManager.transferSnapshot(transferSnapshotProvider, new TranslogTransferListener() {
+                @Override
+
+                public void onUploadComplete(TransferSnapshot transferSnapshot) throws IOException {
+                    transferReleasable.close();
+                    closeFilesIfNoPendingRetentionLocks();
+                    logger.trace("uploaded translog for {} {} ", primaryTerm, generation);
+                }
+
+                @Override
+                public void onUploadFailed(TransferSnapshot transferSnapshot, Exception ex) throws IOException {
+                    transferReleasable.close();
+                    closeFilesIfNoPendingRetentionLocks();
+                }
+            });
+        }
+
+    }
+
+    // Visible for testing
+    public Set<String> allUploaded() {
+        return fileTransferTracker.allUploaded();
+    }
+
+    private boolean syncToDisk() throws IOException {
+        try (ReleasableLock lock = readLock.acquire()) {
+            return current.sync();
+        } catch (final Exception ex) {
+            closeOnTragicEvent(ex);
+            throw ex;
+        }
+    }
+
+    @Override
+    public void sync() throws IOException {
+        try {
+            if (syncToDisk()) {
```

Comment:
As `syncToDisk()` is called before uploading translog files to remote, does it mean `RemoteFsTranslog` has dependency over `LocalTranslog`? Is this a decorator pattern, where we can combine 2 Translog implementations?

Reply:
All the common dependencies are still in `Translog`. Hence `RemoteFsTranslog` doesn't depend upon `LocalTranslog` . You can argue that `RemoteFsTranslog` can extend `LocalTranslog` for now . But we have kept it separate for now as we might plan to not do a local sync at all and directly upload to remote store . And the evolution of  `RemoteFsTranslog` will make it quite different from `LocalTranslog`.